File: crates/harness-subagents/src/internal.rs
//! Internal subagent — wraps a fresh `harness_core::Agent` loop with a
//! restricted system prompt + tool subset and translates its
//! `AgentEvent` stream into [`SubAgentEvent`] frames the outer loop
//! can relay to UIs.
//!
//! Used by `read_doc`, `review`, and the path-A `codex` built-ins.
//! Each one is just a thin factory that picks the right system
//! prompt + tool subset + (optionally) cheaper model; the runner
//! handles the agent-loop plumbing once.
//!
//! The runner does NOT itself install a `with_subagent` scope — the
//! outer agent loop already did when it set up the wrapping tool
//! invocation. That means [`harness_core::emit_subagent`] from
//! inside this runner reaches the outer task-local channel directly,
//! which is exactly what we want (UIs see the inner subagent's
//! events alongside the outer agent's).

use crate::{Artifact, SubAgent, SubAgentInput, SubAgentOutput};
use async_trait::async_trait;
use futures::StreamExt;
use harness_core::{
    emit_subagent, message::Message, Agent, AgentConfig, AgentEvent, BoxError, Conversation,
    LlmProvider, SubAgentEvent, SubAgentFrame, ToolRegistry,
};
use std::sync::Arc;
use tracing::warn;

/// Configuration for an [`InternalSubAgent`]. Built by the per-kind
/// factories (`doc_reader::build`, `reviewer::build`, etc.) and
/// passed straight through to [`InternalSubAgent::new`].
pub struct InternalSubAgentConfig {
    /// Tool name suffix — exposed to the main agent as
    /// `subagent.<name>`. Must be unique within the parent
    /// `SubAgentRegistry`.
    pub name: String,
    /// One-line description shown to the main agent in its tool
    /// catalogue. Must be specific enough that the model knows when
    /// to dispatch to this subagent vs. a sibling.
    pub description: String,
    /// System prompt installed on the inner agent's conversation.
    /// Should constrain the subagent's role + allowed tools + output
    /// shape. The reviewer's prompt for instance pins it to
    /// `requirement.review_verdict` as the only mutation surface.
    pub system_prompt: String,
    /// Inner LLM model identifier. `None` falls back to whatever the
    /// provider's default is.
    pub model: Option<String>,
    /// Hard ceiling on inner agent iterations — protects against
    /// runaway loops without a verification path.
    pub max_iterations: usize,
    /// Provider for the inner agent's LLM calls. Often the same
    /// shared provider as the main agent, but may be a cheaper one
    /// (e.g. doc reader → Haiku).
    pub provider: Arc<dyn LlmProvider>,
    /// ToolRegistry the inner agent gets to use. The composition
    /// root (`apps/jarvis`) builds this with the appropriate subset
    /// — read-only for doc reader / reviewer, full-coding for codex.
    /// Must NOT contain any `subagent.*` tools — recursion is
    /// forbidden in v1.0 to keep the resource budget bounded.
    pub tools: Arc<ToolRegistry>,
    /// Whether the wrapping `subagent.<name>` tool needs human
    /// approval before each invocation. `true` for subagents that
    /// can mutate the workspace (codex / claude_code); `false` for
    /// read-only ones (doc reader / reviewer).
    pub requires_approval: bool,
}

/// Generic Internal subagent runner. Constructors live in the
/// per-kind modules (`doc_reader.rs`, `reviewer.rs`, `codex.rs`) and
/// just pre-fill the config.
pub struct InternalSubAgent {
    config: InternalSubAgentConfig,
}

impl InternalSubAgent {
    pub fn new(config: InternalSubAgentConfig) -> Self {
        Self { config }
    }
}

#[async_trait]
impl SubAgent for InternalSubAgent {
    fn name(&self) -> &str {
        &self.config.name
    }
    fn description(&self) -> &str {
        &self.config.description
    }
    fn input_schema(&self) -> serde_json::Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "task": {
                    "type": "string",
                    "description": "Natural-language task. The subagent's system prompt constrains how it interprets and answers."
                }
            },
            "required": ["task"],
            "additionalProperties": false
        })
    }
    fn requires_approval(&self) -> bool {
        self.config.requires_approval
    }

    async fn invoke(&self, input: SubAgentInput) -> Result<SubAgentOutput, BoxError> {
        let id = uuid::Uuid::new_v4().to_string();
        let push = |event: SubAgentEvent| {
            emit_subagent(SubAgentFrame {
                subagent_id: id.clone(),
                subagent_name: self.config.name.clone(),
                event,
            });
        };

        push(SubAgentEvent::Started {
            task: input.task.clone(),
            model: self.config.model.clone(),
        });

        // Build the inner agent. Conversation seeded with system
        // prompt + the user's task; nothing else.
        let mut conversation = Conversation::new();
        conversation
            .messages
            .push(Message::system(&self.config.system_prompt));
        conversation.messages.push(Message::user(&input.task));

        let mut cfg = AgentConfig::new(
            self.config
                .model
                .clone()
                .unwrap_or_else(|| "default".into()),
        );
        cfg.tools = self.config.tools.clone();
        cfg.max_iterations = self.config.max_iterations;
        // The system prompt is already in the conversation, so don't
        // double-install via AgentConfig.
        cfg.system_prompt = None;
        cfg.session_workspace = Some(input.workspace_root.clone());

        let agent = Arc::new(Agent::new(self.config.provider.clone(), cfg));

        let mut stream = agent.run_stream(conversation);

        let mut final_message = String::new();
        let mut error_message: Option<String> = None;

        while let Some(ev) = stream.next().await {
            match ev {
                AgentEvent::Delta { content } if !content.is_empty() => {
                    push(SubAgentEvent::Delta { text: content });
                }
                AgentEvent::ToolStart {
                    name, arguments, ..
                } => {
                    push(SubAgentEvent::ToolStart { name, arguments });
                }
                AgentEvent::ToolEnd { name, content, .. } => {
                    push(SubAgentEvent::ToolEnd {
                        name,
                        output: content,
                    });
                }
                AgentEvent::Usage { model, usage } => {
                    push(SubAgentEvent::Usage { model, usage });
                }
                AgentEvent::Error { message } => {
                    error_message = Some(message);
                    break;
                }
                AgentEvent::Done { conversation, .. } => {
                    final_message = conversation
                        .last_assistant_text()
                        .map(str::to_owned)
                        .unwrap_or_default();
                    break;
                }
                // Pass through nothing else — inner approval prompts /
                // plan updates / nested subagent events stay private
                // to the inner loop. We don't want the outer UI to
                // see internal-mechanics frames.
                _ => {}
            }
        }

        if let Some(msg) = error_message {
            push(SubAgentEvent::Error {
                message: msg.clone(),
            });
            warn!(subagent = %self.config.name, error = %msg, "subagent failed");
            return Err(format!("subagent error: {msg}").into());
        }

        push(SubAgentEvent::Done {
            final_message: final_message.clone(),
        });

        Ok(SubAgentOutput {
            message: final_message,
            artifacts: extract_artifacts(&self.config.name, &input),
        })
    }
}

/// Per-kind artifact extraction is a future enhancement: the
/// `requirement.review_verdict` tool will record a verdict that we'd
/// surface here as `Artifact::ReviewVerdict`. v1.0 returns no
/// artifacts — the final message is the load-bearing output.
fn extract_artifacts(_name: &str, _input: &SubAgentInput) -> Vec<Artifact> {
    Vec::new()
}

#[cfg(test)]
mod tests {
    use super::*;
    use harness_core::{ChatRequest, ChatResponse, FinishReason};
    use std::path::PathBuf;
    use std::sync::Mutex;

    /// Records every `model` string the runner hands to its provider.
    /// Returns a trivial `ok.` Stop response so the inner agent loop
    /// terminates after one iteration with no tool calls.
    struct ModelRecordingProvider {
        seen_models: Arc<Mutex<Vec<String>>>,
    }

    #[async_trait]
    impl LlmProvider for ModelRecordingProvider {
        async fn complete(&self, req: ChatRequest) -> harness_core::Result<ChatResponse> {
            self.seen_models.lock().unwrap().push(req.model);
            Ok(ChatResponse {
                message: Message::assistant_text("ok."),
                finish_reason: FinishReason::Stop,
                response_id: None,
                usage: None,
            })
        }
    }

    /// Pins the wire contract: whatever `model` the per-kind factory
    /// supplies in `InternalSubAgentConfig` is exactly what the inner
    /// agent puts on `ChatRequest.model` for the LLM provider.
    ///
    /// Regression guard for the runaway-loop incident
    /// (2026-05-08): `subagent.review` was constructed with `model =
    /// None` because no `JARVIS_SUBAGENT_REVIEWER_MODEL` was set and
    /// no `route_policy` had a Review slot. The fallback in this file
    /// pasted the literal `"default"` onto the wire, which Codex
    /// rejected with HTTP 400 every dispatch — and because the work
    /// agent's run was Completed, the picker happily re-picked the
    /// row every tick. The structural fix lives in the composition
    /// root (`apps/jarvis/src/subagents.rs::register_builtins` now
    /// passes the primary model as a universal fallback); this test
    /// catches a regression at this layer if the wiring ever drifts.
    #[tokio::test]
    async fn configured_model_reaches_provider() {
        let seen = Arc::new(Mutex::new(Vec::<String>::new()));
        let provider = Arc::new(ModelRecordingProvider {
            seen_models: seen.clone(),
        });
        let subagent = InternalSubAgent::new(InternalSubAgentConfig {
            name: "test-subagent".into(),
            description: "model wiring test".into(),
            system_prompt: "you are a test stub".into(),
            model: Some("inherited-haiku-test".into()),
            max_iterations: 1,
            provider,
            tools: Arc::new(ToolRegistry::new()),
            requires_approval: false,
        });

        let _ = subagent
            .invoke(SubAgentInput {
                task: "say hi".into(),
                workspace_root: PathBuf::from("/tmp"),
                context: None,
                caller_chain: Vec::new(),
            })
            .await
            .expect("subagent must complete");

        let recorded = seen.lock().unwrap().clone();
        assert!(
            !recorded.is_empty(),
            "provider was not called; runner exited before LLM dispatch"
        );
        assert!(
            recorded.iter().all(|m| m == "inherited-haiku-test"),
            "every ChatRequest must use the configured model; got {recorded:?}"
        );
    }
}

File: packages/subagents/src/internal.ts
// Internal subagent — wraps a fresh agent loop with a restricted system
// prompt + tool subset and translates its AgentEvent stream into
// SubAgentEvent frames the outer loop relays to UIs. Ported from
// harness-subagents/src/internal.rs.
//
// Used by `read_doc`, `review`, and the path-A `codex` built-ins.
//
// Injectable seam: the Rust runner builds a `harness_core::Agent` directly
// from a provider + AgentConfig. Here we take a `createAgent` factory so the
// caller (composition root) owns provider/tool/config construction — the same
// seam the server package uses (`AutoModeDeps.createAgent`). The factory
// returns anything with `runStream(conversation) -> AsyncIterable<AgentEvent>`
// (the @jarvis/core Agent satisfies this).
//
// The runner does NOT itself install a `withSubagent` scope — the outer agent
// loop already did when it set up the wrapping tool invocation, so
// `emitSubagent` from inside this runner reaches the outer channel directly.
import {
  conversationWithSystem,
  userMessage,
  lastAssistantText,
  type AgentEvent,
  type Conversation,
  type JsonValue,
} from "@jarvis/core";
import {
  emitSubagent,
  type Artifact,
  type SubAgent,
  type SubAgentInput,
  type SubAgentOutput,
} from "./subagent.ts";

/**
 * The slice of `@jarvis/core`'s `Agent` the internal runner depends on. A
 * factory returning this is the injectable seam; production wires the real
 * `Agent`, tests pass a canned stub.
 */
export interface SubAgentRunner {
  runStream(conversation: Conversation): AsyncIterable<AgentEvent>;
}

/** Options handed to the `createAgent` factory for one inner run. */
export interface CreateAgentOptions {
  /** Inner LLM model id (already resolved — never the literal "default"). */
  model: string;
  /** Hard ceiling on inner agent iterations. */
  maxIterations: number;
  /** Sandbox root the inner tools resolve under. */
  sessionWorkspace: string;
}

/** Builds an inner agent for one subagent invocation. */
export type CreateAgent = (opts: CreateAgentOptions) => SubAgentRunner;

/**
 * Configuration for an InternalSubAgent. Built by the per-kind factories
 * (`docReader.build`, `reviewer.build`, `codex.build`) and passed straight
 * through to the constructor.
 */
export interface InternalSubAgentConfig {
  /** Tool-name suffix — exposed as `subagent.<name>`. Unique in the registry. */
  name: string;
  /** One-line description for the main agent's tool catalogue. */
  description: string;
  /**
   * System prompt installed on the inner conversation. Constrains the
   * subagent's role + allowed tools + output shape.
   */
  systemPrompt: string;
  /**
   * Inner LLM model id. `undefined` falls back to "default" on the wire — the
   * composition root should always pass a concrete model (see the
   * runaway-loop regression note in internal.test.ts).
   */
  model?: string;
  /** Hard ceiling on inner agent iterations. */
  maxIterations: number;
  /**
   * Factory that builds the inner agent (provider + tools + config). The
   * registry handed to the inner agent MUST NOT contain any `subagent.*`
   * tools — recursion is forbidden in v1.0.
   */
  createAgent: CreateAgent;
  /**
   * Whether the wrapping `subagent.<name>` tool needs approval. true for
   * workspace-mutating subagents (codex / claude_code); false for read-only.
   */
  requiresApproval: boolean;
}

/**
 * Generic internal subagent runner. Per-kind constructors live in the per-kind
 * modules (`doc-reader.ts`, `reviewer.ts`, `codex.ts`) and just pre-fill the
 * config.
 */
export class InternalSubAgent implements SubAgent {
  readonly #config: InternalSubAgentConfig;

  constructor(config: InternalSubAgentConfig) {
    this.#config = config;
  }

  get name(): string {
    return this.#config.name;
  }

  get description(): string {
    return this.#config.description;
  }

  inputSchema(): JsonValue {
    return {
      type: "object",
      properties: {
        task: {
          type: "string",
          description:
            "Natural-language task. The subagent's system prompt constrains how it interprets and answers.",
        },
      },
      required: ["task"],
      additionalProperties: false,
    };
  }

  requiresApproval(): boolean {
    return this.#config.requiresApproval;
  }

  async invoke(input: SubAgentInput): Promise<SubAgentOutput> {
    const id = crypto.randomUUID();
    const push = (event: Parameters<typeof emitSubagent>[0]["event"]): void => {
      emitSubagent({ subagent_id: id, subagent_name: this.#config.name, event });
    };

    const startEvent: { kind: "started"; task: string; model?: string } = {
      kind: "started",
      task: input.task,
    };
    if (this.#config.model !== undefined) startEvent.model = this.#config.model;
    push(startEvent);

    // Inner conversation: system prompt + the user's task; nothing else.
    const conversation: Conversation = conversationWithSystem(this.#config.systemPrompt);
    conversation.messages.push(userMessage(input.task));

    const model = this.#config.model ?? "default";
    const agent = this.#config.createAgent({
      model,
      maxIterations: this.#config.maxIterations,
      sessionWorkspace: input.workspace_root,
    });

    let finalMessage = "";
    let errorMessage: string | undefined;

    for await (const ev of agent.runStream(conversation)) {
      if (ev.type === "delta") {
        if (ev.content.length > 0) push({ kind: "delta", text: ev.content });
      } else if (ev.type === "tool_start") {
        push({ kind: "tool_start", name: ev.name, arguments: ev.arguments });
      } else if (ev.type === "tool_end") {
        push({ kind: "tool_end", name: ev.name, output: ev.content });
      } else if (ev.type === "usage") {
        const { type: _type, model: usageModel, ...usage } = ev;
        push({ kind: "usage", model: usageModel, ...usage });
      } else if (ev.type === "error") {
        errorMessage = ev.message;
        break;
      } else if (ev.type === "done") {
        finalMessage = lastAssistantText(ev.conversation) ?? "";
        break;
      }
      // Pass through nothing else — inner approval prompts / plan updates /
      // nested subagent events stay private to the inner loop.
    }

    if (errorMessage !== undefined) {
      push({ kind: "error", message: errorMessage });
      throw new Error(`subagent error: ${errorMessage}`);
    }

    push({ kind: "done", final_message: finalMessage });

    const artifacts = extractArtifacts();
    const out: SubAgentOutput = { message: finalMessage };
    if (artifacts.length > 0) out.artifacts = artifacts;
    return out;
  }
}

/**
 * Per-kind artifact extraction is a future enhancement (Rust v1.0 returns
 * none; the final message is the load-bearing output).
 */
function extractArtifacts(): Artifact[] {
  return [];
}
